coordinator/src/notifications.rs

```diff
@@ -0,0 +1,356 @@
+            notification_builder.body("Close channel request.");
+            notification_builder.title("Someone wants to close a position with you! 🌻");
+        }
+        NotificationKind::PositionSoonToExpire => {
+            notification_builder.title("Your Position is about to expire");
+            notification_builder.body("Open the app to react.");
+        }
+        NotificationKind::PositionExpired => {
+            notification_builder.title("Your position has expired");
+            notification_builder.body("Open the app to react.");
+        }
+    }
+    notification_builder.finalize()
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct FcmToken(pub String);
+
+impl Display for FcmToken {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", &self.0)
+    }
+}
+
+async fn send_notification<'a>(
+    client: &fcm::Client,
+    api_key: &str,
+    fcm_token: &FcmToken,
+    notification: fcm::Notification<'a>,
+) -> Result<()> {
+    anyhow::ensure!(!api_key.is_empty(), "FCM API key is empty");
+
+    let mut message_builder = fcm::MessageBuilder::new(api_key, &fcm_token.0);
+    message_builder.notification(notification);
+    let message = message_builder.finalize();
+    let response = client
+        .send(message)
+        .await
+        .context("could not send FCM notification")?;
+    tracing::debug!("Sent: {:?}", response);
+    Ok(())
+}
+
+/// Load all recent positions with the DB with associated fcm tokens and send push
+/// notification about expiring/expired positions if needed.
+pub async fn query_and_send_position_notifications(
+    conn: &mut PgConnection,
+    notification_sender: &mpsc::Sender<Notification>,
+) -> Result<()> {
+    let users = user::all(conn)?;
+
+    let positions_with_fcm_tokens = positions::Position::get_all_positions_with_expiry_within(
+        conn,
+        OffsetDateTime::now_utc() - START_OF_EXPIRING_POSITION,
+        OffsetDateTime::now_utc() + END_OF_EXPIRED_POSITION,
+    )?
+    .into_iter()
+    // Join positions with users to add the FCM tokens.
+    // Filter out positions that don't have a FCM token stored in the users
+    // table which is with them.
+    // This can be done at the DB level if it ever becomes a performance issue.
+    .filter_map(|p| {
+        let maybe_fcm_token = users
+            .iter()
+            .find(|u| u.pubkey == p.trader.to_string())
+            .map(|u| FcmToken(u.fcm_token.clone()));
+
+        if let Some(fcm_token) = maybe_fcm_token {
+            Some((p, fcm_token))
+        } else {
+            tracing::warn!(?p, "No FCM token for position");
+            None
+        }
+    })
+    .collect::<Vec<_>>();
+
+    send_expiry_notifications_if_applicable(&positions_with_fcm_tokens, notification_sender).await;
+
+    Ok(())
+}
+
+/// Send notifications to users with positions that are about to expire or have
+/// just expired
+pub async fn send_expiry_notifications_if_applicable(
```

Comment:
Can this be private? It's only being called from the function above.

Reply:
good spotting!